models/src/node/drain.rs

```diff
@@ -0,0 +1,364 @@
+const DELETION_CHECK_INTERVAL: Duration = Duration::from_secs(5);
+// `kubectl drain` by default will wait "forever" for an eviction to complete. We follow suit.
+const DELETION_TIMEOUT: Duration = Duration::from_secs(u64::MAX);
+
+// Some errors while attempting evictions result in retries with exponential backoff.
+// These values configure how long to delay between tries.
+// We should be tenacious in attempting retries, as some workloads are sensitive to being suddenly interrupted.
+const RETRY_BASE_DELAY: Duration = Duration::from_millis(100);
+const RETRY_MAX_DELAY: Duration = Duration::from_secs(20);
+const NUM_RETRIES: usize = 10;
+
+/// Wrapper struct to provide retry configurations for evictions.
+struct RetryStrategy {}
+impl RetryStrategy {
+    fn retry_strategy() -> impl Iterator<Item = Duration> {
+        ExponentialBackoff::from_millis(RETRY_BASE_DELAY.as_millis() as u64)
+            .max_delay(RETRY_MAX_DELAY)
+            .map(jitter)
+            .take(NUM_RETRIES)
+    }
+}
+
+impl tokio_retry::Condition<error::EvictionError> for RetryStrategy {
+    fn should_retry(&mut self, error: &error::EvictionError) -> bool {
+        error.should_retry()
+    }
+}
+
+/// Drains a node of all pods.
+///
+/// The Kubernetes API does not provide an implementation of drain. You must use Pod deletion or the Eviction API
+/// to remove all Pods from a given Node. We opt to use Evictions in order to respect Pod Disruption Budgets.
+///
+/// The implementation of `kubectl drain` can be used as inspiration. While this implementation will halt under
+/// certain special conditions, we have slightly different behavior:
+///
+/// Kubectl by default will not evict nodes under some criteria without further instruction.
+/// By default, we ignore:
+/// * DaemonSet Pods - The DaemonSet controller will not respect node cordons, so we don't battle it.
+/// * Mirror Pods - These are static and cannot be controlled.
+///
+/// Otherwise, we evict pods that kubectl gives special care:
+/// - Pods with local storage that will be deleted when drained (emptyDir).
+/// - Unreplicated pods (Pods without a controller.)
+///
+/// PodDisruptionBudgets can be used to protect these workloads from being unduly interrupted.
+#[instrument(skip(k8s_client), err)]
+pub(crate) async fn drain_node(
+    k8s_client: &kube::Client,
+    node_name: &str,
+) -> Result<(), error::DrainError> {
+    let target_pods = find_target_pods(k8s_client, node_name).await?;
+
+    // Perform the eviction for each pod simultaneously.
+    stream::iter(target_pods)
+        .for_each(move |pod| {
+            let k8s_client = k8s_client.clone();
+            let pod = pod.clone();
+            async move {
+                if evict_pod(&k8s_client, &pod).await.is_ok() {
+                    wait_for_deletion(&k8s_client, &pod).await.ok();
+                }
+            }
+        })
+        .await;
```

Comment:
Could you explain more here?
1. Why do we not care about (not escalate) error::EvictionError in `evict_pod`?
2. How will the error::DrainError in `wait_for_deletion` be caught?

Reply:
There are basically two cases in which evictions don't occur:

1. While attempting to evict, we find that we will never be able to evict. Right now, this happens if the target Pod no longer exists (404) or if we don't have permissions to evict it (403).
2. While attempting to evict, other miscellaneous errors that may be transient occur. (These are all cases that are currently marked as "retriable").


Right now, we allow the software to proceed after emitting a log message in both cases. Though in case 2 we do retries with exponential backoff. I think case 1 we basically always want to treat this way, but an argument could be made for 2 that we may want to bubble all the way back up and provide an error back to the client, forcing them to consider what to do with it. Maybe the client decides to wait and issue its own retry, or even better, cancel its update workflow and allow the brupop controller to attempt to update a different Pod in the meantime.


I think the latter approach I mention above could be a good future improvement, but since we don't currently have a mechanism for the client to effectively "cancel" an update, I'd rather consider that an iterative improvement. Since the number of things that could go wrong to result in us running out of retries is a massive space, I'd prefer for now to drop those errors. I worry that it would be worse for Brupop to frequently become stuck rather than Brupop occasionally failing to gracefully interrupt software under conditions like network failures or Kubernetes bugs.

Reply:
> How will the error::DrainError in wait_for_deletion be caught?

For this error case, we currently also just ignore deletions that don't resolve in time, and allow the node to be updated before the deletion completes. Fortunately, this should never happen, because the default timeout time is MAX_INT, effectively forever.


I'm going to add a section to the README regarding stuck updates, which will mention the cases here where a `drain` can fail to progress.